openssl-dynamic/src/main/c/ssl.c

```diff
@@ -1315,64 +1322,112 @@ TCN_IMPLEMENT_CALL(jobjectArray, SSL, getPeerCertChain)(TCN_STDARGS,
     UNREFERENCED(o);
 
     // Get a stack of all certs in the chain.
-    sk = SSL_get_peer_cert_chain(ssl_);
+#ifdef OPENSSL_IS_BORINGSSL
+    c = tcn_SSL_get_app_data2(ssl_);
+
+    TCN_ASSERT(c != NULL);
+
+    chain = SSL_get0_peer_certificates(ssl_);
+    len = sk_CRYPTO_BUFFER_num(chain);
+
+    if (c->mode == SSL_MODE_SERVER) {
+        // We don't want to include the leaf certificate to mimic the behaviour of SSL_get_peer_cert_chain(...).
+        offset = 1;
+    } else {
+        offset = 0;
+    }
+#else
+    chain = SSL_get_peer_cert_chain(ssl_);
+    len = sk_X509_num(chain);
+    offset = 0;
+#endif // OPENSSL_IS_BORINGSSL
 
-    len = sk_X509_num(sk);
+    len -= offset;
     if (len <= 0) {
         // No peer certificate chain as no auth took place yet, or the auth was not successful.
         return NULL;
     }
+
     // Create the byte[][] array that holds all the certs
     array = (*e)->NewObjectArray(e, len, byteArrayClass, NULL);
 
     for(i = 0; i < len; i++) {
-        cert = sk_X509_value(sk, i);
 
-        buf = NULL;
+#ifdef OPENSSL_IS_BORINGSSL
+        cert = sk_CRYPTO_BUFFER_value(chain, i + offset);
+        length = CRYPTO_BUFFER_len(cert);
+#else
+        cert = sk_X509_value(chain, i + offset);
+
         length = i2d_X509(cert, &buf);
         if (length < 0) {
-            if (buf != NULL) {
-                OPENSSL_free(buf);
-            }
+            OPENSSL_free(buf);
             // In case of error just return an empty byte[][]
             return (*e)->NewObjectArray(e, 0, byteArrayClass, NULL);
         }
+#endif // OPENSSL_IS_BORINGSSL
+
         bArray = (*e)->NewByteArray(e, length);
+
+#ifdef OPENSSL_IS_BORINGSSL
+        (*e)->SetByteArrayRegion(e, bArray, 0, length, (jbyte*) CRYPTO_BUFFER_data(cert));
```

Comment:
There are several instances of this ifdef. You could probably write a single function which takes an `X509` (OpenSSL) / `CRYPTO_BUFFER` (BoringSSL) and calls `SetByteArrayRegion` on the bytes within.

Reply:
good idea will do.